### PR TITLE
Run lint on travis builds and use modern node versions

### PR DESCRIPTION
**Pull request by aaronraimist**
Yes, this will create some noise but maybe that will actually cause someone to fix the lint problems.

### CODE REVIEW

**Comment by aaronraimist** [comment]
If anyone knows why this won't build properly I'd love some help.

It seems to all start going wrong around https://travis-ci.org/vector-im/riot-web/jobs/440888490#L1226
```
ERROR in ./node_modules/matrix-js-sdk/browser-index.js
Module not found: Error: Can't resolve './lib/matrix' in '/home/travis/build/vector-im/riot-web/node_modules/matrix-js-sdk'
 @ ./node_modules/matrix-js-sdk/browser-index.js 1:15-38
 @ ./test/app-tests/joining.js
 @ ./test/app-tests sync \.jsx?$
 @ ./test/all-tests.js
ERROR in ./test/skin-sdk.js
Module not found: Error: Can't resolve 'matrix-react-sdk' in '/home/travis/build/vector-im/riot-web/test'
 @ ./test/skin-sdk.js 9:10-37
 @ ./test/app-tests/joining.js
 @ ./test/app-tests sync \.jsx?$
 @ ./test/all-tests.js
```

Seems like `Setting up matrix-js-sdk` and `Setting up matrix-react-sdk` aren't doing everything they are supposed to do?

File: scripts/generate-eslint-error-ignore-file [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+echo "generating $out"
+
+{
+    cat <<EOF
+# autogenerated file: run scripts/generate-eslint-error-ignore-file to update.
+
+EOF
+
+    ./node_modules/.bin/eslint --no-ignore -f json src test |
+        jq -r '.[] | select((.errorCount) > 0) | .filePath' |
```
Comment (aaronraimist) [inline scripts/generate-eslint-error-ignore-file]: Even though I didn't use this here, I'd lobby for this change (only ignore files with errors, not errors and warnings) to be used here: https://github.com/matrix-org/matrix-react-sdk/blob/87bac56b9c3b0c8688835604a5e51a00220a4e37/scripts/generate-eslint-error-ignore-file#L19

Thoughts?

Comment (aaronraimist) [reply scripts/generate-eslint-error-ignore-file]: Right now lintwithexclusions in matrix-react-sdk makes it look like everything is ok when there are actually `✖ 765 problems (237 errors, 528 warnings)`. This change would at least show a bunch of those warnings so they are more visible.

Comment (aaronraimist) [reply scripts/generate-eslint-error-ignore-file]: Opened https://github.com/matrix-org/matrix-react-sdk/pull/2215

**Comment by turt2live** [comment]
@aaronraimist sorry for missing this. I suspect that fixing the merge conflict and merging in the latest develop would magically fix those errors when you get a chance.

**Comment by aaronraimist** [comment]
@turt2live Still broken. Not sure what the problem is.

To fix the merge conflict, I updated `package-lock.json` by following this advice 
https://github.com/npm/npm/issues/18007#issuecomment-401427694 (running `npm install --package-lock-only`). Not totally sure why that added a bunch of new things.

**Comment by aaronraimist** [comment]
@turt2live so it turns out I can reproduce this locally. It is some kind of issue with `npm link`, related to #7355?

I currently run (fish shell, `; and` is like &&):
```
cd ../matrix-js-sdk/; and npm i; and cd ../matrix-react-sdk/; and npm i; and npm link ../matrix-js-sdk/; and cd ../riot-web/; and npm i; and npm link ../matrix-js-sdk/; and npm link ../matrix-react-sdk/
```
to install all the stuff based on the directions in the README.

After running that, the tests in matrix-react-sdk fail in the same way as they do in the travis build. If I run `npm link ../matrix-js-sdk/`in the `matrix-react-sdk` folder, the tests pass.

I'm hoping https://github.com/npm/rfcs/pull/3 fixes this but is there something else I can do? I'm not sure why this is affecting only my PR.

**Comment by turt2live** [comment]
I'm not too sure at a glance as to what could be going wrong there. I'll run the same tests on my end and see if I can reproduce it, and maybe find out why it is broken.

**Comment by turt2live** [comment]
Alright, so I fixed your original problem by introducing a completely different issue. When I saw this problem while writing https://github.com/matrix-org/matrix-js-sdk/pull/764 an `npm install` fixed it. It's possible something in https://github.com/matrix-org/matrix-js-sdk/pull/764 magically fixes the problem, but it is weird that the js-sdk (and therefore the react-sdk) failed to build at all.

**Comment by aaronraimist** [comment]
:/

**Comment by turt2live** [comment]
Great, so now it is on the original problem:

`Error: Cannot find module 'escope'`

*bashes head against wall*

**Comment by aaronraimist** [comment]
Oh actually I think I can solve that, hang on

**Comment by aaronraimist** [comment]
This PR now relies on  
https://github.com/matrix-org/matrix-react-sdk/pull/2255 and 
https://github.com/matrix-org/matrix-js-sdk/pull/768

**Comment by aaronraimist** [comment]
Travis is going to fail again until those are merged

`fatal: Remote branch lint not found in upstream origin`

**Comment by turt2live** [comment]
Shouldn't https://github.com/vector-im/riot-web/pull/7490/commits/a4ed1afeb5c7f4d239c819ef16e54b5848533056 require a change to the package lock too?

**Comment by aaronraimist** [comment]
Yeah let me delete my package-lock.json and install again so I can commit that. It is currently all screwed up from trying to fix https://github.com/vector-im/riot-web/issues/6694.

**Pull request merged**